validation: stricter internal handling of invalid blocks

mzumsande opened the pull request:
Some fields in validation are set opportunistically by "best effort":
- The `BLOCK_FAILED_CHILD` status (which means that the block index has an invalid predecessor)
- `m_best_header` (the most-work header not known to be invalid).

This means that there are known situations in which these fields are not set when they should be, or set to wrong values. This is tolerated because the fields are not used for anything consensus-critical and triggering these situations involved creating invalid blocks with valid PoW header, so would have a cost attached. Also, having stricter guarantees for these fields requires iterating over the entire block index, which has some DoS potential, especially with any header above the checkpoint being accepted int he past (see e.g. #11531). 

However, there are reasons to change this now:
- RPCs use these fields and can report wrong results
- There is the constant possibility that someone could add code that expects these fields to be correct, especially because it is not well documented that these fields cannot always be relied upon.
- DoS concerns have become less of an issue after #25717 - now an attacker would need to invest much more work because they can't fork off the last checkpoint anymore

This PR continues the work from #30666 to ensure that `BLOCK_FAILED_CHILD` status and `m_best_header` are always correct:
- it adds a call to `InvalidChainFound()` in `AcceptBlock()`.
- it adds checks for `BLOCK_FAILED_CHILD` and `m_best_header`  to `CheckBlockIndex()`. In order to be able to do this, the existing cache in the RPC-only `InvalidateBlock()` is adjusted to handle these as well. These are performance optimizations with the goal of avoiding having a call of `InvalidChainFound()` / looping over the block index after each disconnected block.
I also wrote a fuzz test to find possible edge cases violating `CheckBlockIndex`, which I will PR separately soon.
- it removes the `m_failed_blocks` set, which was a heuristic necessary when we couldn't be sure if a given block index had an invalid predecessor or not. Now that we have that guarantee, the set is no longer needed.

mzumsande committed:
validation: in invalidateblock, calculate m_best_header right away

Before, m_best_header would be calculated only after disconnecting
multiple blocks, letting go of cs_main in the meantime.
This is in preparation for adding checks to CheckBlockIndex()
requiring that m_best_header is the most-work header not known to be invalid.

Co-authored-by: stringintech <stringintech@gmail.com>

mzumsande committed:
validation: in invalidateblock, mark children as invalid right away

Before, they would be marked as invalid only after disconnecting
multiple blocks, letting go of cs_main in the meantime.
This is in preparation for adding a check to
CheckBlockIndex() requiring that descendants of invalid block indexes
are always marked as invalid.
Entries from highpow_outofchain_headers are now only removed if made invalid,
no longer after inserting into setBlockIndexCandidates, because they
might still become invalid later in the second case.
This means that blocks could be inserted multiple times now into
setBlockIndexCandidates - this won't have any effect, so
behavior isn't changed.

mzumsande committed:
validation: call InvalidBlockFound also from AcceptBlock

When a block it found invalid during acceptance (but before connection)
we now mark its descendants with BLOCK_FAILED_CHILD and update
m_best_header when these things weren't done reliably before.

This does not fix a serious bug because the flags and m_best_header were being set on a best-effort basis before
and not used for anything critical.
Setting these reliably has a slight performance cost (iterating over the
entire block index) but leads to more consistency in validation and allows removing m_failed_blocks in a later commit.
This can only be triggered by providing a block with sufficient PoW
that is otherwise invalid, so it is not a DoS vector.

mzumsande committed:
validation: remove m_failed_blocks

After changes in previous commits, we now mark all blocks that descend from an invalid block
immediately as the block is found invalid. This happens both in the AcceptBlock
and ConnectBlock stages of block validation.
As a result, the pindexPrev->nStatus check in AcceptBlockHeader is now sufficient to detect
invalid blocks and checking m_failed_blocks there is no longer necessary.

mzumsande committed:
validation: Add more checks to CheckBlockIndex()

This adds checks that
1) Descendants of invalid block indexes are also marked invalid
2) m_best_header cannot be invalid, and there can be no valid
block with more work than it.

stratospher reviewed (commented):
ACK b2136da9

- Went through the logs in `rpc_invalidateblock.py` and checked how the children were marked invalid/`m_best_header` was set on master vs in this PR
- Went through logs of some functional tests like `feature_block.py` and verified that BLOCK_FAILED_CHILD wasn't being set in [`m_failed_blocks` loop](https://github.com/bitcoin/bitcoin/blob/35bf426e02210c1bbb04926f4ca2e0285fbfcd11/src/validation.cpp#L4395)
- Ran `invalidateblock` RPC on a full node.

mzumsande commented on src/validation.cpp:
```diff
@@ -5367,6 +5359,8 @@ void ChainstateManager::CheckBlockIndex()
         if (pindexFirstInvalid == nullptr) {
             // Checks for not-invalid blocks.
             assert((pindex->nStatus & BLOCK_FAILED_MASK) == 0); // The failed mask cannot be set for blocks without invalid parents.
+        } else {
```
I changed the comment to "Invalid blocks and their descendants must be marked as invalid" - doesn't really make sense to me to make an exception for the block itself.

mzumsande commented on src/validation.cpp:
```diff
@@ -3729,6 +3739,28 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+        // Mark descendants of the invalidated block as invalid
+        auto range = cand_invalid_descendants.equal_range(invalid_walk_tip);
+        for (auto it = range.first; it != range.second; ++it) {
+            it->second->nStatus |= BLOCK_FAILED_CHILD;
```
Done, made it a bit shorter:
`it->second->nStatus = (it->second->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;`

stratospher reviewed (commented):
reACK 4ba2e48.

stickies-v commented:
PSA: Bitcoin Core PR Review Club will cover this PR in its next meeting at 2025-03-05 at 17:00 UTC. See https://bitcoincore.reviews/31405 for notes, questions, and instructions on [how to join](https://bitcoincore.reviews/).

stickies-v commented on src/validation.cpp:
```diff
@@ -3691,6 +3690,11 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+    // Maps fork block in chain to the candidates for invalidation.
+    std::multimap<const CBlockIndex*, CBlockIndex*> cand_invalid_descendants;
+    // Map to cache candidates for m_best_header
+    std::multimap<const arith_uint256, CBlockIndex*> best_header_blocks_by_work;
```
Since this map is a superset of `candidate_blocks_by_work `, I think we could just have one, and instead check for `IsValid(BLOCK_VALID_TRANSACTIONS) && HaveNumChainTxs()` further down when we update `setBlockIndexCandidates`? I suspect the memory impllications should be negligible either way, but I think it would clean up the code a bit?

E.g. something like:

<details>
<summary>git diff on 4ba2e480ff</summary>

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index f9c900ef27..ead0454f7b 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3689,12 +3689,10 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
     // To avoid walking the block index repeatedly in search of candidates,
     // build a map once so that we can look up candidate blocks by chain
     // work as we go.
-    std::multimap<const arith_uint256, CBlockIndex *> candidate_blocks_by_work;
+    std::multimap<const arith_uint256, CBlockIndex *> highpow_outofchain_headers;
     // Map to cache candidates not in the main chain that might need invalidating.
     // Maps fork block in chain to the candidates for invalidation.
     std::multimap<const CBlockIndex*, CBlockIndex*> cand_invalid_descendants;
-    // Map to cache candidates for m_best_header
-    std::multimap<const arith_uint256, CBlockIndex*> best_header_blocks_by_work;
 
     {
         LOCK(cs_main);
@@ -3706,16 +3704,9 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
             // Instead, consider only non-active-chain blocks that have at
             // least as much work as where we expect the new tip to end up.
             if (!m_chain.Contains(candidate) &&
-                    !CBlockIndexWorkComparator()(candidate, pindex->pprev) &&
-                    candidate->IsValid(BLOCK_VALID_TRANSACTIONS) &&
-                    candidate->HaveNumChainTxs()) {
-                candidate_blocks_by_work.insert(std::make_pair(candidate->nChainWork, candidate));
-            }
-            // Similarly, populate cache for blocks not in main chain to invalidate
-            if (!m_chain.Contains(candidate) &&
-                !CBlockIndexWorkComparator()(candidate, pindex->pprev)) {
+                    !CBlockIndexWorkComparator()(candidate, pindex->pprev)) {
+                highpow_outofchain_headers.insert(std::make_pair(candidate->nChainWork, candidate));
                 cand_invalid_descendants.insert(std::make_pair(m_chain.FindFork(candidate), candidate));
-                best_header_blocks_by_work.insert(std::make_pair(candidate->nChainWork, candidate));
             }
         }
     }
@@ -3774,11 +3765,9 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
         // Determine new best header
         if (m_chainman.m_best_header->nStatus & BLOCK_FAILED_MASK) {
             m_chainman.m_best_header = invalid_walk_tip->pprev;
-            auto best_header_it = best_header_blocks_by_work.lower_bound(m_chainman.m_best_header->nChainWork);
-            while (best_header_it != best_header_blocks_by_work.end()) {
-                if (best_header_it->second->nStatus & BLOCK_FAILED_MASK) {
-                    best_header_it = best_header_blocks_by_work.erase(best_header_it);
-                } else {
+            auto best_header_it = highpow_outofchain_headers.lower_bound(m_chainman.m_best_header->nChainWork);
+            while (best_header_it != highpow_outofchain_headers.end()) {
+                if (!(best_header_it->second->nStatus & BLOCK_FAILED_MASK)) {
                     if (!CBlockIndexWorkComparator()(best_header_it->second, m_chainman.m_best_header)) {
                         m_chainman.m_best_header = best_header_it->second;
                     }
@@ -3788,11 +3777,12 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
         }
 
         // Add any equal or more work headers to setBlockIndexCandidates
-        auto candidate_it = candidate_blocks_by_work.lower_bound(invalid_walk_tip->pprev->nChainWork);
-        while (candidate_it != candidate_blocks_by_work.end()) {
-            if (!CBlockIndexWorkComparator()(candidate_it->second, invalid_walk_tip->pprev)) {
-                setBlockIndexCandidates.insert(candidate_it->second);
-                candidate_it = candidate_blocks_by_work.erase(candidate_it);
+        auto candidate_it = highpow_outofchain_headers.lower_bound(invalid_walk_tip->pprev->nChainWork);
+        while (candidate_it != highpow_outofchain_headers.end()) {
+            if (!CBlockIndexWorkComparator()(candidate_it->second, invalid_walk_tip->pprev) &&
+                candidate_it->second->IsValid(BLOCK_VALID_TRANSACTIONS) &&
+                candidate_it->second->HaveNumChainTxs()) {
+                    setBlockIndexCandidates.insert(candidate_it->second);
             } else {
                 ++candidate_it;
             }

```
</details>

mzumsande replied:
That's a good idea! I think we can even go one step further and use `highpow_outofchain_headers` for everything, also for `cand_invalid_descendants` (at the cost of doing a few `GetAncestor()` calls, which should be cheap enough). I pushed an updated version that uses this approach.

stickies-v reviewed (commented):
Concept ACK for making `m_best_header` and `nStatus` more reliably correct, and the trade-offs seem reasonable. Even though the diff is small, I find reasoning about these changes fairly complex and require a lot of context, so I'll need to think through the implications more.

mzumsande commented on src/validation.cpp:
```diff
@@ -3754,6 +3764,29 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
             m_blockman.m_dirty_blockindex.insert(to_mark_failed);
         }
 
+        // Mark descendants of the invalidated block as invalid
+        // (possibly replacing a pre-existing BLOCK_FAILED_VALID with BLOCK_FAILED_CHILD)
+        auto range = cand_invalid_descendants.equal_range(invalid_walk_tip);
+        for (auto it = range.first; it != range.second; ++it) {
+            it->second->nStatus = (it->second->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
+        }
```
There is some more discussion about this in #31835, where I wrote my opinion. Since `BLOCK_FAILED_CHILD` has never been used for anything, it's hard to say what the intentions really were, but my impression is that the first invalid block should be marked as `BLOCK_FAILED_VALID` and all of its descendants as `BLOCK_FAILED_CHILD` - so having both `BLOCK_FAILED_VALID` and  `BLOCK_FAILED_CHILD` at the same time doesn't make sense to me.

mzumsande commented on src/validation.cpp:
```diff
@@ -3707,6 +3711,12 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+            // Similarly, populate cache for blocks not in main chain to invalidate
+            if (!m_chain.Contains(candidate) &&
+                !CBlockIndexWorkComparator()(candidate, pindex->pprev)) {
+                cand_invalid_descendants.insert(std::make_pair(m_chain.FindFork(candidate), candidate));
```
With `pindex->pprev` being the predecessor of the block passed to the `invalidateblock` rpc call, the logic in `InvalidateBlock()` is to go from the tip back to `pindex->pprev` -  and we may need to update other forked descendants of the blocks encountered along that way (which must have more accumulated work than their parents), which is why we put them in the cache. But all blocks with lower work can never be encountered by that process and therefore don't need to be in the caches.
I'll add more documentation with my next push.

stickies-v replied:
Oh sorry, I get it now, thanks. For some reason (I think I had just been reviewing the next section dealing with disconnecting the tip) I was confusing `pindex` for the chain tip here. I don't really have any doc suggestions, so can be marked as resolved.

stringintech commented:
Concept ACK.

I wrote two test cases to understand the general behavior of `InvalidateBlock`, not specifically to test this PR's changes:
```
A->B->C->D->E->F (main chain)
A->B->C->G (fork 1)
A->B->H->I->J (fork 2)
A->B->H->K (fork 3)
```

In the [first test](https://github.com/stringintech/bitcoin/blob/1b36dd6a66a44e8756d92e71baa7676646b5f1ce/src/test/validation_block_tests.cpp#L455), I invalidate block C (on active chain) which should invalidate C, D, E, F, and G. The [second test](https://github.com/stringintech/bitcoin/blob/1b36dd6a66a44e8756d92e71baa7676646b5f1ce/src/test/validation_block_tests.cpp#L509) invalidates H (on a fork) which should invalidate H, I, J, and K.

Both tests behave the same on master and PR branches. The first test fails on both because D, E, and F are not correctly marked as BLOCK_FAILED_CHILD.

I have two questions:

1 )
I found what appears to be a bug in `InvalidateBlock`:
https://github.com/bitcoin/bitcoin/blob/4637cb1eec48d1af8d23eeae1bb4c6f8de55eed9/src/validation.cpp#L3750

The condition seems incorrect and should probably be:
```cpp
if (to_mark_failed->pprev == invalid_walk_tip ...) {
```
The assertions that fail in the first test are for blocks D, E, and F which were all part of the best chain before invalidation, which led me to this part of the code.

2 )
To validate my understanding: Since both branches behave similarly (even on master the best header and failed child statuses are set correctly - except for the ones not set as child failed because of the bug), is it correct that the main goal of this PR is to maintain proper block statuses and best header between tip disconnections by marking children as invalid right away and recalculating the best header immediately rather than doing it later?

mzumsande commented:
> Update: Just saw the PR https://github.com/bitcoin/bitcoin/pull/31835 which fixes this issue. :)

Heh, looks like you are at least the third person who found that bug independently, first one was in #16856!

>  Since both branches behave similarly (even on master the m_best_header and BLOCK_FAILED_CHILD statuses are set correctly because of the final call to InvalidChainFound - except for the ones not set as BLOCK_FAILED_CHILD because of the potential bug), is it correct that the main goal of changes applied in InvalidateBlock is to maintain proper block statuses and best header between tip disconnections by marking children as invalid right away and recalculating the best header immediately rather than doing it later?

Yes, that's correct. The main goal of this PR to add these assumptions to `CheckBlockIndex()` - this means they should always be true when `cs_main` is not held. Since `InvalidateBlock` will release `cs_main` and call `CheckBlockIndex()` after each step, some tests would fail without the changes to `InvalidateBlock()`  - even if the final `InvalidChainFound()` in that function will eventually correct it!

mzumsande committed:
validation: cache all headers with enough PoW in invalidateblock

We now include blocks without HaveNumChainTxs() / lower validation status
than BLOCK_VALID_TRANSACTIONS. These checks are still performed at the
spot where we use the cache to insert into setBlockIndexCandidates.

This is in preparation for using the cache for more things than
just setBlockIndexCandidates candidates in the following commits.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

mzumsande commented:
[4ba2e48](https://github.com/bitcoin/bitcoin/commit/4ba2e480ffa0b77113953bee4ff5c9349e277e7e) to [fd39a64](https://github.com/bitcoin/bitcoin/commit/fd39a6420dad4cfc8bfff73f32df663da1ec00e2):
I changed the approach to `InvalidateBlock()` after the suggestion by @stickies-v: Instead of having 3 slightly different caches, now only one cache is used for everything (`highpow_outofchain_headers`). This should make the changes more simple and have a negligible overhead of a few additional `GetAncestor()` calls.

I also added a call to CheckBlockIndex after each iteration of `InvalidateBlock()` (which releases `cs_main`, so any other thread could call CBI, which means that the block index cannot be in an inconsistent state, even temporarily). This should make it clear why the changes to `InvalidateBlock()` are necessary, even if (as @stringintech noted) the additional checks to `CheckBlockIndex` might not have led to an failure otherwise.

stringintech commented on src/validation.cpp:
```diff
@@ -3754,14 +3753,34 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
-            } else {
-                ++candidate_it;
+            }
+            ++candidate_it;
+        }
+
+        // Determine new best header if that was the invalidated block or one of its descendants
+        if (m_chainman.m_best_header->nStatus & BLOCK_FAILED_MASK) {
+            m_chainman.m_best_header = invalid_walk_tip->pprev; // this block is definitely still valid at this point, but there may be better ones
+            auto best_header_it = highpow_outofchain_headers.lower_bound(m_chainman.m_best_header->nChainWork);
+            while (best_header_it != highpow_outofchain_headers.end()) {
+                if (!CBlockIndexWorkComparator()(best_header_it->second, m_chainman.m_best_header)) {
+                    m_chainman.m_best_header = best_header_it->second;
+                }
+                ++best_header_it;
             }
         }
```
Can the `m_best_header` update logic be merged into the while loop that handles the `nStatus` updates? Not sure if it is necessarily better; but I was thinking of something like:

Adding this just before the while loop:
```c++
const bool best_header_needs_update = m_chainman.m_best_header->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip;
if (best_header_needs_update) {
  m_chainman.m_best_header = invalid_walk_tip->pprev;
}
``` 

And then here:
```suggestion
            if (best_header_needs_update &&
                !CBlockIndexWorkComparator()(candidate_it->second, m_chainman.m_best_header)) {
                m_chainman.m_best_header = candidate_it->second;
            }
            ++candidate_it;
        }
```

mzumsande replied:
Yes, I like that solution, it's less code and since we would `continue`  when encountering an invalid header before we'd never set an invalid block as best header, so it's correct. 
I added you as co-author of that commit with the email from your github profile - let me know if you want met to change anything about that!

stringintech replied:
Thanks a lot! Yes the email is okay.

stringintech commented on src/validation.cpp:
```diff
@@ -3754,14 +3753,34 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+        while (candidate_it != highpow_outofchain_headers.end()) {
+            if (candidate_it->second->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip) {
+                candidate_it->second->nStatus = (candidate_it->second->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
+                // If invalidated, the block is irrelevant for setBlockIndexCandidates and m_best_header and can be removed fro mthe cache
```
Small typo at the end of the comment

mzumsande replied:
fixed

mzumsande commented on src/validation.cpp:
```diff
@@ -3754,14 +3753,34 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+        auto candidate_it = highpow_outofchain_headers.lower_bound(invalid_walk_tip->pprev->nChainWork);
+        while (candidate_it != highpow_outofchain_headers.end()) {
+            if (candidate_it->second->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip) {
+                candidate_it->second->nStatus = (candidate_it->second->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
```
yes added - nothing bad would happen if we didn't (because BLOCK_FAILED_CHILD would be recalculated at next startup) but I think it's definitely better to do it.

sedited reviewed (approved):
crACK 7b6fe5a2196d53a3024dfe7f11b316a50acc3cb7

This reduces complexity overall and I am happy that the slightly verbose block in `AcceptBlockHeader` and the `m_dirty_blockindex` are gone. It is not really straight forward to review though and I hope I did not miss any edge cases or real performance drawbacks. Might be good to get some more eyes on this. Having a better abstraction on the BlockMap / block tree data structure that just handles all of this internally could make things easier in the future.

mzumsande commented on src/validation.cpp:
```diff
@@ -5434,6 +5435,8 @@ void ChainstateManager::CheckBlockIndex()
             // Checks for not-invalid blocks.
             assert((pindex->nStatus & BLOCK_FAILED_MASK) == 0); // The failed mask cannot be set for blocks without invalid parents.
+        } else {
+            assert((pindex->nStatus & BLOCK_FAILED_MASK)); // Invalid blocks and their descendants must be marked as invalid
```
I missed this before, but it's fixed now!

mzumsande commented:
[7b6fe5a](https://github.com/bitcoin/bitcoin/commit/7b6fe5a2196d53a3024dfe7f11b316a50acc3cb7) to [96e6371](https://github.com/bitcoin/bitcoin/commit/96e6371328c7715ef814da5831d08b5d30285dd3):
Addressed https://github.com/bitcoin/bitcoin/pull/31405#discussion_r2019820910 and removed unnecessary braces.

sedited reviewed (approved):
Re-ACK 96e6371328c7715ef814da5831d08b5d30285dd3

stratospher reviewed (approved):
reACK 96e6371.

I liked the new approach of using 1 cache for everything and the smaller diff. Also went through the logs in rpc_invalidate.py to make sure that everything works as intended.

mzumsande commented:
[96e6371](https://github.com/bitcoin/bitcoin/commit/96e6371328c7715ef814da5831d08b5d30285dd3) to [41860d9](https://github.com/bitcoin/bitcoin/commit/41860d98408f52c29e78f7708e8a0355ea5cc04c): rebased due to conflict with https://github.com/bitcoin/bitcoin/pull/31835

stickies-v reviewed (commented):
The implementation simplifications since my last review are great (and require a small update to the PR description which still mentions the addition of two caches).

I'm still not 100% comfortable that the changes don't introduce any unwanted side-effects, so I want to give it another full review first. Approach ACK.

mzumsande commented on src/validation.cpp:
```diff
@@ -3716,6 +3716,7 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
 
         // Make sure the queue of validation callbacks doesn't grow unboundedly.
         if (m_chainman.m_options.signals) LimitValidationInterfaceQueue(*m_chainman.m_options.signals);
+        m_chainman.CheckBlockIndex();
```
We don't "need" the check - but what we want to guarantee is that whenever `cs_main` is not held, the block index must be in a consistent state (i.e. `CheckBlockIndex()` would pass) - because at this point, other threads can change the block index / call CBI, leading to issues such as #16444.
This spot is pretty critical, because even though `InvalidateBlock` will clean up in the end, it releases `cs_main` for stability reasons multiple times (so that the node doesn't become unresponsive for too long) - so having the BlockIndex in a consistent state at this spot is the reason for the changes to  `InvalidateBlock`  in this PR, so I think it's helpful to also check for this on regtest.

(I'll address the rest of the feedback soon, thanks!)

stickies-v replied:
Thank you, I think I understand now. The commit message and the placement of the new `CheckBlockIndex()` call in 79bef1bc52afab860782d6d9a6016b93f94c4503 being before (rather than after) the block index mutations led me to believe that we were worried about _other_ threads leaving the block index in an inconsistent state, and made this particular addition feel rather arbitrary.
Instead, it seems you want to ensure that `InvalidateBlock()` does not leave the block index in an inconsistent state whenever `cs_main` is released.

If that's correct, then I think:
- `CheckBlockIndex()` should be called after, rather than before, logic that mutates the block index. Of course, in a loop (and with the extra call after the loop finishes), this is largely (but not entirely) equivalent, so it's mostly about readability and signaling intent.
  - One concern here is that the loop could (and sometimes does*) return early, in which case we might never call `CheckBlockIndex()`, which should probably be addressed with an `RAII` struct that does the check whenever it goes out of scope? Or alternatively, just add a `CheckBlockIndex()` call whenever we return early.
- `CheckBlockIndex()` should be called right before releasing, but **while it still has** the `cs_main` lock, to avoid synchronization issues (effectively making `CheckBlockIndex()` an `EXCLUSIVE_LOCKS_REQUIRED(::cs_main)` method)

\* Note: the current implementation in 79bef1bc52afab860782d6d9a6016b93f94c4503 will not call `CheckBlockIndex()` if `DisconnectTip()` [fails](https://github.com/bitcoin/bitcoin/pull/31405/commits/79bef1bc52afab860782d6d9a6016b93f94c4503#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98L3738). At first sight, it doesn't look like `DisconnectTip()` can bring `m_block_index` in an inconsistent state, so probably fine?

mzumsande replied:
> If that's correct, then I think:

Yes, that's correct.

> CheckBlockIndex() should be called right before releasing, but while it still has the cs_main lock, to avoid synchronization issues (effectively making CheckBlockIndex() an EXCLUSIVE_LOCKS_REQUIRED(::cs_main) method)

`CheckBlockIndex()` is called from multiple places in validation, some of which hold `cs_main`, some of which don't.
My understanding is that it must always pass if `cs_main` isn't held (because at that time, other threads could change the block index / invoke CBI themselves), but not necessarily while `cs_main`  is held  (not everything is atomic).

In general, I think that CBI calls should be strategically placed to spots where (in case of bugs) those are most likely to occur - no need to cover every return path etc.

On second thought, I think the added call (even if it helped find issues during development) might be overkill, considering that we already have a call [here](https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/src/validation.cpp#L3738C16-L3738C31). So I have remove the commit that adds it for now.

stickies-v replied:
> So I have remove the commit that adds it for now.

That seems good to me.

> My understanding is that it must always pass if cs_main isn't held (because at that time, other threads could change the block index / invoke CBI themselves), but not necessarily while cs_main  is held (not everything is atomic).

This still feels awkward to me, like we're sprinkling tests throughout the code in case they catch something without a clear aim. If we're trying to ensure that `InvalidateBlock()` does not leave the block index (BI) in an inconsistent state, then releasing `cs_main` so _another thread can change the BI_ seems like the test isn't just testing `InvalidateBlock()`, but also (especially) those other threads? If we want to test `InvalidateBlock()`, then imo we _must_ call CBI before releasing `cs_main`. And if other threads might be problematic, then we should add CBI call(s) in those locations?

Anyway, commit removed, so this can be resolved now. Thanks!

stickies-v replied:
> So I have remove the commit that adds it for now.

Actually, it looks like 79bef1bc52afab860782d6d9a6016b93f94c4503 is still here (with HEAD 9d001dc1b1627f7a973e21227955bba1f473d6aa)

mzumsande replied:
> Actually, it looks like https://github.com/bitcoin/bitcoin/commit/79bef1bc52afab860782d6d9a6016b93f94c4503 is still here

oops, fixed

> releasing cs_main so another thread can change the BI

I wasn't suggesting this, the reason why we are releasing cs_main frequently during invalidateblock in the first place is so that the node/GUI doesn't become unresponsive in case many blocks in a row are disconnected.

In the past, there was a problem with the intermediate state in `InvalidateBlock()` (fixed in https://github.com/bitcoin/bitcoin/pull/16849) after cs_main was released that was only discovered because in a functional test  it happened that headers were received/CBI was invoked in a different thread (https://github.com/bitcoin/bitcoin/issues/16444.).

stickies-v replied:
Can be marked as resolved, thanks! I think we were talking past each other a bit but resolved offline.

mzumsande commented on src/validation.cpp:
```diff
@@ -3771,6 +3778,10 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
                 setBlockIndexCandidates.insert(candidate_it->second);
             }
+            if (best_header_needs_update &&
+                !CBlockIndexWorkComparator()(candidate_it->second, m_chainman.m_best_header)) {
```
`CBlockIndexWorkComparator()(a, b))` means that a is strictly smaller than b.
`!CBlockIndexWorkComparator()(b, a))` means that a is smaller or or equal to b (in this case, equal means the same block index because of the [pointer address criterion](https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/src/node/blockstorage.cpp#L165-L168)).

I think that's why sometimes one or the other is used.

In this case, it doesn't really matter because the two elements are chosen such that they cannot be the same, so I'll change it to the simpler version.

stickies-v replied:
> I think that's why sometimes one or the other is used.

Right, that makes sense.

I think in the two other cases where `CBlockIndexWorkComparator` is used in this PR, comparing directly on `nChainwork` would be more readable and more correct?
- https://github.com/bitcoin/bitcoin/pull/31405/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R3704: "that have at least as much work as where we expect the new tip to end up."
- https://github.com/bitcoin/bitcoin/pull/31405/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R3778: the `setBlockIndexCandidates` doc states "entries that have as much work as our current tip or more"

<details>
<summary>git diff on 9d001dc1b1</summary>

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index b7fe228bbf..b4f6bb40f4 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3701,7 +3701,7 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
             // Instead, consider only non-active-chain blocks that have at
             // least as much work as where we expect the new tip to end up.
             if (!m_chain.Contains(candidate) &&
-                !CBlockIndexWorkComparator()(candidate, pindex->pprev) &&
+                (candidate->nChainWork >= pindex->pprev->nChainWork) &&
                 !(candidate->nStatus & BLOCK_FAILED_MASK)) {
                 highpow_outofchain_headers.insert(std::make_pair(candidate->nChainWork, candidate));
             }
@@ -3775,7 +3775,7 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
                 candidate_it = highpow_outofchain_headers.erase(candidate_it);
                 continue;
             }
-            if (!CBlockIndexWorkComparator()(candidate, invalid_walk_tip->pprev) &&
+            if ((candidate->nChainWork >= invalid_walk_tip->pprev->nChainWork) &&
                 candidate->IsValid(BLOCK_VALID_TRANSACTIONS) &&
                 candidate->HaveNumChainTxs()) {
                 setBlockIndexCandidates.insert(candidate);

```
</details>


Where it is [used](https://github.com/bitcoin/bitcoin/pull/31405/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R3784) to update `m_best_header`, which is just a single block index, it seems sticking to `CBlockIndexWorkComparator` seems appropriate for determinism.

mzumsande replied:
I think that this wouldn't be correct: We use `CBlockIndexWorkComparator` during normal adding to `setBlockIndexCandidates` ([here](https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/src/validation.cpp#L3840C41-L3840C79)) and also use it while asserting that blocks that sort worse are not in the set [here](https://github.com/bitcoin/bitcoin/blob/baa848b8d38187ce6b24a57cfadf1ea180209711/src/validation.cpp#L5412). If we'd insert blocks in the set that have the same `nChainWork` but sort worse (due to `nSequence`), I would expect the assert to be hit.

That being said, some of the comments about this (e.g. also in `TryAddBlockIndexCandidate`) are misleading in my opinion. I have reworded the spot in InvalidateBlock, any other spots should be cleaned up separately.

mzumsande commented on src/validation.cpp:
```diff
@@ -3757,12 +3757,19 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+
+        const bool best_header_needs_update{m_chainman.m_best_header->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip};
+        if (best_header_needs_update) {
+            m_chainman.m_best_header = invalid_walk_tip->pprev; // this block is definitely still valid at this point, but there may be better ones
```
done

mzumsande commented on src/validation.cpp:
```diff
@@ -3757,12 +3757,19 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
         // Mark descendants of the invalidated block as invalid
         // (possibly replacing a pre-existing BLOCK_FAILED_VALID with BLOCK_FAILED_CHILD)
         // Add any equal or more work headers that are not invalidated to setBlockIndexCandidates
+        // Recalculate the best header if it became invalid.
```
done

mzumsande commented on src/validation.cpp:
```diff
@@ -3754,17 +3754,24 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
         auto candidate_it = highpow_outofchain_headers.lower_bound(invalid_walk_tip->pprev->nChainWork);
         while (candidate_it != highpow_outofchain_headers.end()) {
+            if (candidate_it->second->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip) {
+                candidate_it->second->nStatus = (candidate_it->second->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
```
done. The suggested comment is a bit too strong for my taste: In reality the distinction doesn't matter, even if both were set nothing bad would happen, see #31835. This is more for being consistent with how it's handled elsewhere.

stickies-v replied:
Thanks, the new comment "Children of failed blocks should be marked as BLOCK_FAILED_CHILD instead." works for me, can be marked as resolved.

mzumsande commented on src/validation.cpp:
```diff
@@ -3754,15 +3753,35 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
-        auto candidate_it = candidate_blocks_by_work.lower_bound(invalid_walk_tip->pprev->nChainWork);
-        while (candidate_it != candidate_blocks_by_work.end()) {
-            if (!CBlockIndexWorkComparator()(candidate_it->second, invalid_walk_tip->pprev)) {
+        // Mark descendants of the invalidated block as invalid
```
done

stickies-v reviewed (approved):
ACK ed172d3002da68a3ddbd5d13e7d3f0618c1378d4

I've thoroughly gone through this PR many times now, and I don't think I can assure myself anymore than I am now that it is correct and safe. Still, the code touched in this PR is fairly complicated and coupled, and I am not as familiar with this validation logic as others, so my main concern is that there are unexpected side-effects I've not considered.

Nits can be ignored, comments (if correct) can be addressed in other pulls.

sedited reviewed (approved):
Re-ACK ed172d3002da68a3ddbd5d13e7d3f0618c1378d4

achow101 commented:
ACK ed172d3002da68a3ddbd5d13e7d3f0618c1378d4

Would prefer if others more familiar with consensus could review this before merging.

ryanofsky reviewed (approved):
Code review ACK ed172d3002da68a3ddbd5d13e7d3f0618c1378d4. Seems a lot nicer to track more reliable information instead of relying on best-effort heuristics if this is now possible.

This change looks ready to merge with all the reviews and discussion so far, but there are some suggestions from 3 weeks ago https://github.com/bitcoin/bitcoin/pull/31405#pullrequestreview-2843060953 that could be responded to. Could merge this or hold off depending on your preference @mzumsande

mzumsande commented:
> This change looks ready to merge with all the reviews and discussion so far, but there are some suggestions from 3 weeks ago [#31405 (review)](https://github.com/bitcoin/bitcoin/pull/31405#pullrequestreview-2843060953) that could be responded to. Could merge this or hold off depending on your preference @mzumsande

Please hold off merging, I'll address @stickies-v  and your comments later this week!

mzumsande committed:
doc: Improve m_best_header documentation

mzumsande commented on src/validation.cpp:
```diff
@@ -3699,13 +3698,12 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
-                candidate_blocks_by_work.insert(std::make_pair(candidate->nChainWork, candidate));
+                !CBlockIndexWorkComparator()(candidate, pindex->pprev) &&
+                !(candidate->nStatus & BLOCK_FAILED_MASK)) {
+                highpow_outofchain_headers.insert(std::make_pair(candidate->nChainWork, candidate));
```
done

mzumsande commented on src/validation.cpp:
```diff
@@ -3755,11 +3754,14 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+                candidate->IsValid(BLOCK_VALID_TRANSACTIONS) &&
+                candidate->HaveNumChainTxs()) {
+                setBlockIndexCandidates.insert(candidate);
+                candidate_it = highpow_outofchain_headers.erase(candidate_it);
```
good point - yes, that comment belonged to the following commit. I moved it there in a slightly shorter form.

mzumsande commented on src/validation.cpp:
```diff
@@ -3754,15 +3752,38 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+                setBlockIndexCandidates.insert(candidate);
+            }
+            if (best_header_needs_update &&
+                CBlockIndexWorkComparator()(m_chainman.m_best_header, candidate)) {
```
I think that this behavior is because the existing behavior during normal addition ([here](https://github.com/bitcoin/bitcoin/blob/ae024137bda9fe189f4e7ccf26dbaffd44cbbeb6/src/node/blockstorage.cpp#L230)), is to set it to the header received first, and it is 
not changed in `ReceivedBlockTransactions` based on which full blocks are first received in full/connected.

So in this scenario
1.) Receive Header 1
2.) Receive Header 2 of same height
3.) Receive/Connect Block 2
4.) Receive Block 1
the tip will be at Block 2, but `m_best_header` will be at header 1.

This is why changing the `CheckBlockIndex` assert to use `CBlockIndexWorkComparator` should lead to assert failures. This could be changed in a follow-up, but we should check more closely if this would be beneficial to any of the uses of `m_best_header` (maybe not?).

In this PR, I changed the `InvalidateBlock` code to use `nChainWork` to be similar to the way it's done in RecalculateBlockHeader (to not raise the expectation that it needs to fulfill CBlockIndexWorkComparator ordering) - although it's arbitrary which block will be picked in case of multiple candidates. 
I also added a comment to `m_best_header` that it does not follow `CBlockIndexWorkComparator` tiebreak scores.

mzumsande commented on src/validation.cpp:
```diff
@@ -3754,15 +3752,38 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+        const bool best_header_needs_update{m_chainman.m_best_header->GetAncestor(invalid_walk_tip->nHeight) == invalid_walk_tip};
+        if (best_header_needs_update) {
+            m_chainman.m_best_header = invalid_walk_tip->pprev; // pprev is definitely still valid at this point, but there may be better ones
+        }
```
fixed

mzumsande commented on src/validation.cpp:
```diff
@@ -4540,6 +4540,7 @@ bool ChainstateManager::AcceptBlock(const std::shared_ptr<const CBlock>& pblock,
         if (state.IsInvalid() && state.GetResult() != BlockValidationResult::BLOCK_MUTATED) {
             pindex->nStatus |= BLOCK_FAILED_VALID;
             m_blockman.m_dirty_blockindex.insert(pindex);
+            ActiveChainstate().InvalidChainFound(pindex);
```
I changed the commit message based on your suggestion. I dropped the "make the code less complicated" part - 
it's a bit subjective and anyway not my main motivation for this PR:
I think that in an area as central as validation we should avoid having structures that are populated on a "best-effort" basis (and therefore cannot be subjected to rigorous checks as in `CheckBlockIndex`) and trust that no one will ever rely on them for anything too serious in the future.

mzumsande commented on src/validation.cpp:
```diff
@@ -4540,6 +4540,7 @@ bool ChainstateManager::AcceptBlock(const std::shared_ptr<const CBlock>& pblock,
         if (state.IsInvalid() && state.GetResult() != BlockValidationResult::BLOCK_MUTATED) {
             pindex->nStatus |= BLOCK_FAILED_VALID;
             m_blockman.m_dirty_blockindex.insert(pindex);
+            ActiveChainstate().InvalidChainFound(pindex);
```
> which is referring to the loop in Chainstate::SetBlockFailureFlags

It's referring to `SetBlockFailureFlags` and `RecalculateBestHeader` which both loop over the block index.

> unnecessary database writes

I agree that this is a potential follow-up (to #31835, pinging @stratospher). It's probably not performance-critical, because it's also guarded by PoW (creating multiple PoW-valid headers that build on a failed block is very costly).

stickies-v commented on src/validation.cpp:
```diff
@@ -3754,15 +3752,42 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
+                candidate->IsValid(BLOCK_VALID_TRANSACTIONS) &&
+                candidate->HaveNumChainTxs()) {
+                setBlockIndexCandidates.insert(candidate);
+                // Do not remove candidate from the highpow cache, because it might be a descendant of the block being invalidated
```
nit for greppability (and line lenght)
```suggestion
                // Do not remove candidate from highpow_outofchain_headers, because it might be a
                // descendant of the block being invalidated which needs to be marked failed later.
```

stickies-v reviewed (approved):
re-ACK fdcdc2cdeadb5d36076f1b94b54261e22e031354

Mostly documentation changes addressing review feedback. Non-trivial code changes are:
- [deduplicating](https://github.com/bitcoin/bitcoin/compare/ed172d3002da68a3ddbd5d13e7d3f0618c1378d4..1f09e3754363382806cc1e3d6bed25cd235ac5f4#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R4523-R4525) `InvalidChainFound()` logic, addressing https://github.com/bitcoin/bitcoin/pull/31405#discussion_r2090791378
- [Choosing](https://github.com/bitcoin/bitcoin/compare/ed172d3002da68a3ddbd5d13e7d3f0618c1378d4..1f09e3754363382806cc1e3d6bed25cd235ac5f4#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R3787) new `m_best_header` solely based on `nChainwork`, addressing https://github.com/bitcoin/bitcoin/pull/31405#discussion_r2090961175

ryanofsky commented on src/validation.h:
```diff
@@ -1038,7 +1038,10 @@ class ChainstateManager
+    /** Best header we've seen so far for which the block is not known to be invalid
+        (used, among others, for getheaders queries' starting points).
+        In case of multiple best headers with the same work, it could point to any:
+        CBlockIndexWorkComparator tiebreaker rules don't apply here. */
```
In commit "doc: Improve m_best_header documentation" (fdcdc2cdeadb5d36076f1b94b54261e22e031354)

It seems like it would be nice to use CBlockIndexWorkComparator consistently for comparisons, but since it's probably beyond scope of this PR, it's good to have this documentation comment mentioning the lack of consistency.

Just in case its useful for future reference, the post describing the current situation in the most detail is https://github.com/bitcoin/bitcoin/pull/31405#discussion_r2129780775

mzumsande commented on src/validation.cpp:
```diff
@@ -4537,9 +4520,9 @@ bool ChainstateManager::AcceptBlock(const std::shared_ptr<const CBlock>& pblock,
-            pindex->nStatus |= BLOCK_FAILED_VALID;
-            m_blockman.m_dirty_blockindex.insert(pindex);
+        Assume(state.IsInvalid());
+        if (state.IsInvalid()) {
```
done

mzumsande commented on src/validation.h:
```diff
@@ -1038,28 +1038,10 @@ class ChainstateManager
-    /** Best header we've seen so far (used for getheaders queries' starting points). */
+    /** Best header we've seen so far for which the block is not known to be invalid
+        (used, among others, for getheaders queries' starting points).
+        In case of multiple best headers with the same work, it could point to any:
```
changed to "because" - it was meant to imply a causality between the sentences.

mzumsande commented:
[fdcdc2c](https://github.com/bitcoin/bitcoin/commit/fdcdc2cdeadb5d36076f1b94b54261e22e031354) to [76c835f](https://github.com/bitcoin/bitcoin/commit/76c835f77205d04c4206acad1a8a636562c22f79): addressed nits

ryanofsky reviewed (approved):
Code review ACK fdcdc2cdeadb5d36076f1b94b54261e22e031354. Looks good, thanks for many new clarifications! Only changes since last review were switching from a CBlockIndexWorkComparator to a nChainWork comparison, switching from InvalidChainFound to InvalidBlockFound without changing behavior, and improving many comment. I left more suggestions below which are not important and all only about comments.

ryanofsky reviewed (approved):
Code review ACK f6b782f3aad4a6bcf823a9a0fabb4418bca1eea1 with only minor code & comment updates

stickies-v commented:
re-ACK f6b782f3aad4a6bcf823a9a0fabb4418bca1eea1

No changes except minor ones addressing review feedback wrt documentation updates and inlining.

sedited reviewed (approved):
Re-ACK f6b782f3aad4a6bcf823a9a0fabb4418bca1eea1

ryanofsky reviewed (approved):
This seems ready to merge, but it's also a subtle change so it could be helpful if more reviewers want to take a look and ack. I'm planning to merge this in a day or two if nothing new comes up.
